File: 525-contiguous-array/contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        diff_idx = {}
        ones = zeros = 0
        res = 0

        for i, n in enumerate(nums):
            if n == 0:
                zeros += 1
            else:
                ones += 1
            
            diff = ones - zeros
            if diff not in diff_idx:
                diff_idx[diff] = i
            
            if diff == 0:
                res = ones + zeros
            else:
                idx = diff_idx[diff]
                res = max(res, i - idx)
        return res

File: 525-contiguous-array/contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size(), res = 0, count = 0;
        vector<int> diff(2 * n + 1, INT_MIN);
        diff[n] = -1;

        for (int i = 0; i < n; ++i) {
            count += nums[i] == 1 ? 1 : -1;
            if (diff[count + n] >= -1) {
                res = max(res, i - diff[count + n]);
            } else {
                diff[count + n] = i;
            }
        }

        return res;
    }
};